trinity/utils/db.py

```diff
@@ -1,35 +1,40 @@
-class MemoryDB:
-    kv_store = None
+from typing import TypeVar, Generic, Dict
 
-    def __init__(self, kv_store=None):
+K = TypeVar('K')
+V = TypeVar('V')
+
+class MemoryDB(Generic[K, V]):
```

Comment:
@pipermerriam @gsalgado I went all in with typings here and made this generic. Is this how you would like to have it or would you rather want this stuff to be non-generic and annotated with `object`? Personally, I'd vote for making it generic but basically because I'm a fan of Generics in other languages. I'm a bloody Python newbie and don't know about the implications that may have.

Reply:
I think type annotations should be as specific as possible, so here it'd be preferable to use `class MemoryDB(Mapping[bytes, bytes]):`. Although I've just noticed trinity's MemoryDB is not used anywhere, so we should probably remove it

Reply:
oh. So the `MemoryDB` sounded like something generic to me where the type of the `key` and `value` may depend on the use case. If you pin it down to `[bytes, bytes]` we'll only be able to use it as such (which may be fine if it's really only meant to be used in that way) but with `K` and `V` being generic one could use a `MemoryDB` somewhere where the key is a `str` and the` value` is a `byte` whereas somewhere else it would be used with both `key` and `value` being of type `byte` for instance. All that while still maintaining full type safety for both use cases.

But if the `MemoryDB` isn't used anyway, I'll just remove it in a separate commit.

Reply:
Because of python's dynamic typing nature, any container type will be _generic_, which can be problematic if the actual implementation cannot handle elements of _all_ possible types. That's one of the cases where type hints are useful, as they help ensure those containers are only used with types they know how to handle. Also, at this point the type hints are not used to provide actual type safety -- they're only used by static analysis tools like mypy. You might want to read https://www.python.org/dev/peps/pep-0484/ to understand the background and goals for type hints

Reply:
Yes, I get that. That's the same with TypeScript which just adds type hints on top of JavaScript. But still `mypy` (or `tsc` in the TypeScript case) will enforce the type constraints at **development time** and save you tons of trouble at least. This already gives you much more safety.

Let's imagine you have a `Stack`. If you don't make the `Stack` generic you either:

1. Lose type safety and allow for this to happen

```
my_stack.push('foo')
my_stack.push(3)
```

Which will may blow up at runtime if the purpose of this stack instance was to hold only strings.

or

2. You have to create a `StringStack`, `ByteStack`, `IntStack` etc to get type safety at the cost of code duplication. This is essentially what you are doing when you say you'd would like to fix `MemoryDB` to `[byte, byte]`.

But with generics you can make a `Stack<T>` (sorry, using TypeScript syntax for my personal convenience :sweat_smile: :)

and then you can have many different stacks that are - **at development time** - constraint to be type checked.

So the following would be disallowed by mypy/tsc.

```
my_stack = new Stack<string>();
my_stack.push('foo')
// blows up at *development time*
my_stack.push(3)
```

Still you can use your `Stack<T>` to make another *instance* of the stack that is strongly typed to only serve ints.

```
my_stack = new Stack<int>();
my_stack.push(3)
```

That's the beauty of generics which are possible to enforce at *development time* with Python 3.6 type hints.

Reply:
We've chosen to lean towards the *strict* side of things.  Our database objects like the one in question will always be `bytes -> bytes`.

Reply:
Ok, no problem. But do we agree that I should simply nuke the `MemoryDB` in a separate commit since it isn't used at all across the code base.

Reply:
Ok, I noticed that even though `MemoryDB` isn't used in trinity yet, there's at least a test that uses it so I brought it back and added the `[bytes, bytes]` type hints you suggested.